Accept Array-like objects seamlessly in builtins

hubertp opened the pull request:
### Pull Request Description

Most of the time, rather than defining the type of the parameter of the builtin, we want to accept every Array-like object i.e. Vector, Array, polyglot Array etc.
Rather than writing all possible combinations, and likely causing bugs on the way anyway as we already saw, one should use `CoerceArrayNode` to convert to Java's `Object[]`.
Added various test cases to illustrate the problem.

### Checklist

Please include the following checklist in your PR:

- [ ] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

hubertp committed:
Accept Array-like objects seamlesly in builtins

Most of the time, rather than defining the type of the parameter of the
builtin, we want to accept every Array-like object i.e. Vector, Array,
polyglot Array etc.
Rather than writing all possible combinations, and likely causing bugs
on the way anyway as we already saw, one should use `CoerceArrayNode`
to convert to Java's `Object[]`.
Added various test cases to illustrate the problem.

hubertp committed:
update changelog

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/mutable/CoerceArrayNode.java:
```diff
@@ -0,0 +1,85 @@
+    }
+  }
+
+  @ExplodeLoop
```
Will exploding loop work? The `argsLength` would have to be a compile time constant then...

hubertp replied:
True

JaroslavTulach commented on test/Tests/src/Data/Polyglot_Spec.enso:
```diff
@@ -0,0 +1,26 @@
+        Polyglot.invoke poly_date "atStartOfDay" [] . should_equal date
+        Polyglot.invoke poly_date "atStartOfDay" [].to_array . should_equal date
+
+    Test.specify "should be able to crate a new polyglot object using the constructor" <|
```
```suggestion
    Test.specify "should be able to create a new polyglot object using the constructor" <|
```

JaroslavTulach reviewed (approved):
Great.

hubertp committed:
PR review

hubertp committed:
Update test/Tests/src/Data/Polyglot_Spec.enso

Co-authored-by: Jaroslav Tulach <jaroslav.tulach@enso.org>

hubertp committed:
nits

radeusgd commented on test/Tests/src/Semantic/Meta_Spec.enso:
```diff
@@ -148,7 +153,7 @@ spec = Test.group "Meta-Value Manipulation" <|
 
     Test.specify "should allow to get the source location of a frame" pending=location_pending <|
         src = Meta.get_source_location 0
-        loc = "Meta_Spec.enso:150:15-40"
+        loc = "Meta_Spec.enso:155:15-40"
```
Can we move this test to a separate file? Then we won't need to update it on each change.

hubertp replied:
Done

hubertp committed:
Refactor location dependent test

Moved location-dependent test cases out of Meta spec. That way we don't
have to update them every single time we add new test to Meta.

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/mutable/CoerceArrayNode.java:
```diff
@@ -0,0 +1,84 @@
+      Builtins builtins = Context.get(this).getBuiltins();
+      Atom err = builtins.error().makeTypeError(builtins.array(), arr, "arr");
+      throw new PanicException(err, this);
+
```
```suggestion
```

hubertp committed:
Update engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/mutable/CoerceArrayNode.java

Co-authored-by: Radosław Waśko <radoslaw.wasko@enso.org>
